File: agentmem/sdk/typescript/src/webhooks.ts
/**
 * Webhook receiver utilities for AgentMem.
 *
 * Every AgentMem webhook POST includes an `X-Lians-Signature` header of the
 * form `sha256=<hex>`.  Use `verifyWebhookSignature` to authenticate the
 * request before processing the payload.
 *
 * @example
 * // Express handler
 * import { verifyWebhookSignature, WebhookPayload } from "lians/webhooks";
 *
 * app.post("/webhooks/agentmem", express.raw({ type: "application/json" }), (req, res) => {
 *   const sig = req.headers["x-agentmem-signature"] as string;
 *   if (!verifyWebhookSignature(req.body, sig, process.env.AGENTMEM_WEBHOOK_SECRET!)) {
 *     return res.status(401).json({ error: "Invalid signature" });
 *   }
 *   const event: WebhookPayload = JSON.parse(req.body.toString());
 *   // handle event.event, event.data ...
 *   res.sendStatus(200);
 * });
 */

import { createHmac, timingSafeEqual } from "crypto";
import type { WebhookPayload, WebhookEventType } from "./types.js";

export type { WebhookPayload, WebhookEventType };

/**
 * Verify the HMAC-SHA256 signature on an incoming webhook.
 *
 * @param body    - Raw request body as a Buffer or UTF-8 string
 * @param header  - Value of the `X-Lians-Signature` header (e.g. `sha256=abc123…`)
 * @param secret  - The webhook secret returned when the endpoint was registered
 * @returns true if the signature is valid, false otherwise
 */
export function verifyWebhookSignature(
  body: Buffer | string,
  header: string,
  secret: string,
): boolean {
  if (!header.startsWith("sha256=")) return false;
  const expected = "sha256=" + createHmac("sha256", secret)
    .update(typeof body === "string" ? body : body)
    .digest("hex");
  try {
    return timingSafeEqual(Buffer.from(header), Buffer.from(expected));
  } catch {
    return false;
  }
}

/**
 * Parse and validate a raw webhook body into a typed payload.
 * Throws if the signature is invalid or the body is not valid JSON.
 *
 * @param body    - Raw request body as a Buffer or UTF-8 string
 * @param header  - Value of the `X-Lians-Signature` header
 * @param secret  - Webhook secret
 */
export function parseWebhookPayload<T = Record<string, unknown>>(
  body: Buffer | string,
  header: string,
  secret: string,
): WebhookPayload<T> {
  if (!verifyWebhookSignature(body, header, secret)) {
    throw new Error("AgentMem webhook signature verification failed");
  }
  const text = typeof body === "string" ? body : body.toString("utf8");
  return JSON.parse(text) as WebhookPayload<T>;
}

File: sdk/python/src/lians/webhooks.py
"""
AgentMem Python SDK — webhook signature verification.

Every AgentMem webhook delivery is HMAC-SHA256-signed.
The signature is in the ``X-AgentMem-Signature`` header as ``sha256=<hex>``.

Usage::

    from lians.webhooks import verify_webhook_signature, parse_webhook_payload

    @app.post("/webhook")
    async def handle(request: Request):
        body = await request.body()
        header = request.headers.get("X-AgentMem-Signature", "")
        if not verify_webhook_signature(body, header, WEBHOOK_SECRET):
            raise HTTPException(status_code=401)
        payload = parse_webhook_payload(body, header, WEBHOOK_SECRET)
        ...
"""
from __future__ import annotations

import hashlib
import hmac
import json
from typing import Any


def verify_webhook_signature(body: bytes, signature_header: str, secret: str) -> bool:
    """
    Verify the HMAC-SHA256 signature on an AgentMem webhook delivery.

    Args:
        body:              Raw request body bytes.
        signature_header:  Value of the ``X-AgentMem-Signature`` header,
                           e.g. ``sha256=abc123...``
        secret:            The webhook secret returned at endpoint registration.

    Returns:
        True if the signature is valid, False otherwise.

    The comparison is constant-time to prevent timing attacks.
    """
    if not signature_header.startswith("sha256="):
        return False
    received = signature_header[len("sha256="):]
    expected = hmac.new(secret.encode(), body, hashlib.sha256).hexdigest()
    return hmac.compare_digest(expected, received)


def parse_webhook_payload(
    body: bytes,
    signature_header: str,
    secret: str,
) -> dict[str, Any]:
    """
    Verify signature and parse webhook payload as a dict.

    Raises:
        ValueError: if signature verification fails.
        json.JSONDecodeError: if the body is not valid JSON.
    """
    if not verify_webhook_signature(body, signature_header, secret):
        raise ValueError(
            "Webhook signature verification failed. "
            "Ensure the secret matches the one returned at registration."
        )
    return json.loads(body)
